File: src/main/java/dev/foxgirl/torrent/bencode/BencodeList.java
package dev.foxgirl.torrent.bencode;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public final class BencodeList extends ArrayList<BencodeElement> implements BencodeElement {

    public BencodeList() {
        super();
    }

    public BencodeList(int capacity) {
        super(capacity);
    }

    public BencodeList(@NotNull List<? extends BencodeElement> source) {
        super(Objects.requireNonNull(source, "Argument 'source'"));
    }

    public void addInteger(long value) {
        add(BencodeInteger.of(value));
    }
    public void addInteger(int index, long value) {
        add(index, BencodeInteger.of(value));
    }

    public void addString(@NotNull String value) {
        add(BencodeString.of(value));
    }
    public void addString(int index, @NotNull String value) {
        add(index, BencodeString.of(value));
    }

    public void setInteger(int index, long value) {
        set(index, BencodeInteger.of(value));
    }

    public void setString(int index, @NotNull String value) {
        set(index, BencodeString.of(value));
    }

    @Override
    public @NotNull BencodeType getType() {
        return BencodeType.LIST;
    }

    @Override
    public @NotNull BencodeList copy() {
        var list = new BencodeList(size());
        for (var element : this) {
            list.add(element.copy());
        }
        return list;
    }

    @Override
    public @NotNull BencodeList asList() {
        return this;
    }

    @Override
    public @NotNull String toString() {
        return super.toString();
    }
}

File: src/main/kotlin/dev/foxgirl/torrent/bencode2/BencodeList.kt
package dev.foxgirl.torrent.bencode2

/** Represents a Bencode list. */
class BencodeList : ArrayList<Bencode>, MutableList<Bencode>, Bencode {

    constructor() : super()
    constructor(capacity: Int) : super(capacity)
    constructor(collection: Collection<Bencode>) : super(collection)

    override fun add(element: Bencode) = super.add(element)
    fun add(element: Int) = add(Bencode.of(element))
    fun add(element: Long) = add(Bencode.of(element))
    fun add(element: String) = add(Bencode.of(element))

    override fun add(index: Int, element: Bencode) = super.add(index, element)
    fun add(index: Int, element: Int) = add(index, Bencode.of(element))
    fun add(index: Int, element: Long) = add(index, Bencode.of(element))
    fun add(index: Int, element: String) = add(index, Bencode.of(element))

    override operator fun set(index: Int, element: Bencode) = super.set(index, element)
    operator fun set(index: Int, element: Int) = set(index, Bencode.of(element))
    operator fun set(index: Int, element: Long) = set(index, Bencode.of(element))
    operator fun set(index: Int, element: String) = set(index, Bencode.of(element))

    override fun copy(): BencodeList {
        val list = BencodeList(size)
        for (element in this) {
            list.add(element.copy())
        }
        return list
    }

}